python/metatensor-learn/metatensor/learn/nn/equivariant_transform.py

```diff
@@ -0,0 +1,145 @@
+from typing import List, Optional, Union
+
+import torch
+from torch.nn import Module
+
+from .._backend import Labels, TensorMap
+from ._utils import _check_module_map_parameter
+from .module_map import ModuleMap
+
+
+class EquivariantTransform(Module):
+    """
+    A custom :py:class:`torch.nn.Module` that applies an arbitrary shape- and
+    equivariance-preserving transformation to an input :py:class:`TensorMap`.
+
+    ``module`` is passed as a callable with parameters ``in_features`` and optionally
+    ``dtype`` and ``device``. This callable constructs an arbitrary shape-preserving
+    :py:class:`torch.nn.Module` transformation (i.e. :py:class:`torch.nn.Tanh`).
+    Separate instantiations are created for each block using the metadata information
+    passed in ``in_keys`` and ``in_features``.
+
+    For invariant blocks in `in_keys` and indexed by `invariant_key_idxs`, the
+    transformation is applied as is. For covariant blocks, an invariant multiplier that
+    preserves covariance is created from the transformation.
+
+    Each parameter can be passed as a single value of its expected type, which is used
+    as the parameter for all blocks. Alternatively, they can be passed as a list to
+    control the parameters applied to each block indexed by the keys in :param in_keys:.
+
+    :param module: :py:class:`callable`, the callable to apply to the invariant blocks
+        and to the invariant built from the covariant blocks.
+    :param in_keys: :py:class:`Labels`, the keys that are assumed to be in the input
+        tensor map in the :py:meth:`forward` method.
+    :param invariant_key_idxs: list of int, the indices of the invariant keys present in
+        `in_keys` in the input :py:class:`TensorMap`. Only blocks for these keys will
+        have bias applied according to the user choice. Covariant blocks will not have
+        bias applied.
+    :param in_features: list of int, the number of features in the input tensor for each
+        block indexed by the keys in :param in_keys:. If passed as a single value, the
+        same number of features is assumed for all blocks.
+    :param out_properties: list of :py:class`Labels` (optional), the properties labels
+        of the output. By default the output properties are relabeled using
+        Labels.range.
+    :param device: :py:class:`str` or :py:class:`torch.device`, the computational device
+        to use for calculations.
+    :param dtype: :py:class:`torch.dtype` , the scalar type to use to store parameters.
+    """
+
+    def __init__(
+        self,
+        module: callable,
```

Comment:
Should this be

```suggestion
        module: torch.nn.Module,
```

Reply:
So the idea was to not have it yet initialised, so that different instances for each block can be created on runtime according to feature size.

For instance, `module` could just be i.e. a `Tanh` activation, but also any shape-preserving transformation like:
```py
def module(in_features: int, dtype, device):
    return torch.nn.Sequential(
        torch.nn.Linear(in_features, 10),
        torch.nn.Tanh(),
        torch.nn.Linear(10, in_features),
    )
```

with the only requirement being that it is parametrized by `in_features`. Do you see a better way of doing it?

Reply:
This is deviating a bit from usual torch API, so I would rather not do something like this. How about taking either a single module (then used by all blocks, with weight sharing) or a list of modules (one per block)?

Also, as a side comment, why does this need to preserve the number of features?

Reply:
So the shape needs to be preserved so that it can be used as an element wise multiplier for covariants, and for that reason we can't weight share as the dimensions for all blocks are in general different.

And yes I suppose we could pass a list